### PR TITLE
feat: Upgrade to ens-contracts@1.6.0 with enhanced ETH registrar controller

### PR DESCRIPTION
**Pull request by storywithoutend**
## Overview

This PR upgrades ENSjs to use **@ensdomains/ens-contracts@1.6.0**, introducing a new unwrapped ETH registrar controller that supports records, reverse records, and referral tracking while maintaining backward compatibility with legacy contracts.

**Key improvements:**
- 🎯 **Enhanced registration**: New controller supports setting records and reverse records at registration time
- 🔗 **Referral tracking**: Built-in support for tracking registration sources via referrer parameter  
- 🧹 **Code cleanup**: Removed unused tuple-based registration types and legacy code
- 🔧 **Improved testing**: Added deployment scripts and generators for comprehensive testing
- 🐛 **DNS fixes**: RFC 8484 compliant DNS-over-HTTPS implementation

---

### New Primary Controller (Unwrapped/Legacy Names)
The main upgrade introduces a new `ETHRegistrarController` that mints traditional (unwrapped) ENS names with enhanced capabilities:

**Mainnet:**
- **New controller**: [`0x59E16fcCd424Cc24e280Be16E11Bcd56fb0CE547`](https://github.com/ensdomains/ensjs/blob/6dff985a/packages/ensjs/src/contracts/consts.ts#L47-L49)
- **New resolver**: [`0xF29100983E058B709F3D539b0c765937B804AC15`](https://github.com/ensdomains/ensjs/blob/6dff985a/packages/ensjs/src/contracts/consts.ts#L53-L55)
- **New bulk renewal**: [`0xc649947a460B135e6B9a70Ee2FB429aDBB529290`](https://github.com/ensdomains/ensjs/blob/6dff985a/packages/ensjs/src/contracts/consts.ts#L38-L40)

**Sepolia:**
- **New controller**: [`0xfb3cE5D01e0f33f41DbB39035dB9745962F1f968`](https://github.com/ensdomains/ensjs/blob/6dff985a/packages/ensjs/src/contracts/consts.ts#L97-L99)
- **New resolver**: [`0xC8D9Af801CbFB8AdFADe3CB5E25F88A5C010B44C`](https://github.com/ensdomains/ensjs/blob/6dff985a/packages/ensjs/src/contracts/consts.ts#L104-L106)
- **New bulk renewal**: [`0x6394b694a8C0DC716e447802E568F0Fb2c4E0965`](https://github.com/ensdomains/ensjs/blob/6dff985a/packages/ensjs/src/contracts/consts.ts#L88-L90)

### Wrapped Controller Support
Added explicit addresses for the wrapped controller (NameWrapper-based NFT names):

**Mainnet:**
- **Wrapped controller**: [`0x253553366Da8546fC250F225fe3d25d0C782303b`](https://github.com/ensdomains/ensjs/blob/6dff985a/packages/ensjs/src/contracts/consts.ts#L68-L70) 
- **Wrapped resolver**: [`0x231b0Ee14048e9dCcD1d247744d114a4EB5E8E63`](https://github.com/ensdomains/ensjs/blob/6dff985a/packages/ensjs/src/contracts/consts.ts#L71-L73)
- **Wrapped bulk renewal**: [`0xa12159e5131b1eEf6B4857EEE3e1954744b5033A`](https://github.com/ensdomains/ensjs/blob/6dff985a/packages/ensjs/src/contracts/consts.ts#L74-L76)

**Sepolia:**
- **Wrapped controller**: [`0x4477cAc137F3353Ca35060E01E5aEb777a1Ca01B`](https://github.com/ensdomains/ensjs/blob/6dff985a/packages/ensjs/src/contracts/consts.ts#L119-L121)
- **Wrapped resolver**: [`0x8948458626811dd0c23EB25Cc74291247077cC51`](https://github.com/ensdomains/ensjs/blob/6dff985a/packages/ensjs/src/contracts/consts.ts#L122-L124)
- **Wrapped bulk renewal**: [`0x4EF77b90762Eddb33C8Eba5B5a19558DaE53D7a1`](https://github.com/ensdomains/ensjs/blob/6dff985a/packages/ensjs/src/contracts/consts.ts#L125-L127)

### Legacy Contract References
Previous controller addresses are preserved as `legacyEthRegistrarController` and `legacyPublicResolver` for operations on older registrations:
- [Legacy controller mapping](https://github.com/ensdomains/ensjs/blob/6dff985a/packages/ensjs/src/contracts/consts.ts#L77-L79) (mainnet)
- [Legacy resolver mapping](https://github.com/ensdomains/ensjs/blob/6dff985a/packages/ensjs/src/contracts/consts.ts#L80-L82) (mainnet)

### Network Support Changes
- ❌ **Removed Holesky support** - ENS is deprecating Holesky testnet ([consts.ts#L10](https://github.com/ensdomains/ensjs/blob/6dff985a/packages/ensjs/src/contracts/consts.ts#L10))
- ✅ **Mainnet + Sepolia** remain fully supported

---

### New Registration Capabilities

The updated registration functions now support:

1. **Records at registration time** - Set text records, addresses, and content hash in the same transaction
2. **Reverse record setup** - Configure primary name (reverse record) during registration  
3. **Referral tracking** - Include a referrer hash to track registration sources

**Updated function signatures:**
- [`commitName()`](https://github.com/ensdomains/ensjs/blob/6dff985a/packages/ensjs/src/functions/wallet/commitName.ts#L20-L22) - Now accepts `referrer` parameter
- [`registerName()`](https://github.com/ensdomains/ensjs/blob/6dff985a/packages/ensjs/src/functions/wallet/registerName.ts#L19-L22) - Enhanced with records and referral support
- [`renewNames()`](https://github.com/ensdomains/ensjs/blob/6dff985a/packages/ensjs/src/functions/wallet/renewNames.ts) - Updated for new bulk renewal contract

### Contract ABI Updates

Updated ABIs to match ens-contracts@1.6.0:
- [ETH Registrar Controller ABI](https://github.com/ensdomains/ensjs/blob/6dff985a/packages/ensjs/src/contracts/ethRegistrarController.ts#L2-L102) - New error types: `CommitmentNotFound`, enhanced error details with timestamps
- [Commitment structure](https://github.com/ensdomains/ensjs/blob/6dff985a/packages/ensjs/src/contracts/ethRegistrarController.ts#L264-L276) - Now includes `referrer` as final parameter
- [Bulk Renewal ABI](https://github.com/ensdomains/ensjs/blob/6dff985a/packages/ensjs/src/contracts/bulkRenewal.ts#L2-L100) - Updated for new contract

---

### Removed Legacy Registration Code

**Deleted files:**
- [`legacyCommitName.ts`](https://github.com/ensdomains/ensjs/blob/be91390b/packages/ensjs/src/functions/wallet/legacyCommitName.ts) (127 lines) - Replaced by main `commitName()` 
- [`legacyRegisterName.ts`](https://github.com/ensdomains/ensjs/blob/be91390b/packages/ensjs/src/functions/wallet/legacyRegisterName.ts) (159 lines) - Replaced by main `registerName()`
- [`legacyRegisterHelpers.ts`](https://github.com/ensdomains/ensjs/blob/be91390b/packages/ensjs/src/utils/legacyRegisterHelpers.ts) (147 lines) - Helpers no longer needed
- All associated test files

**Why removed:** These "legacy" functions were actually for the *previous* new controller. With ens-contracts@1.6.0, the main functions now handle all cases, eliminating the need for separate legacy variants.

### Removed Unused Tuple-Based Types

**Removed from [`registerHelpers.ts`](https://github.com/ensdomains/ensjs/blob/6dff985a/packages/ensjs/src/utils/registerHelpers.ts):**
- `CommitmentTuple` type (lines 58-67) 
- `RegistrationTuple` type (lines 69-78)
- `makeCommitmentTuple()` function
- `makeRegistrationTuple()` function  
- `makeCommitmentFromTuple()` function
- `makeCommitment2()` duplicate function

**Verified safe to remove:**
- ✅ Not used anywhere in ensjs-v3 codebase
- ✅ Not imported by ens-app-v3 (checked 5 files that import from ensjs)
- ✅ Replaced by CallData-based approach (`RegistrationCallData`, `makeCommitmentFromCallData`)

**Kept types** (actively used):
- ✅ `RegistrationParameters` - Used in 5 files in ens-app-v3
- ✅ `ReverseRecordParameter` - Used in referrable name generator
- ✅ `RegistrationCallData` - Internal type for contract calls

Updated exports in [`utils/index.ts`](https://github.com/ensdomains/ensjs/blob/6dff985a/packages/ensjs/src/utils/index.ts#L130-L138) to remove unused exports.

---

### New Deployment Scripts

**Referrable name generator** ([`deploy/00_register_referrable.ts`](https://github.com/ensdomains/ensjs/blob/d53216e3/packages/ensjs/deploy/00_register_referrable.ts)):
- Utility for testing referral tracking functionality
- Generates test names with various referrer configurations
- Helper script: [`utils/referrableNameGenerator.ts`](https://github.com/ensdomains/ensjs/blob/d53216e3/packages/ensjs/utils/referrableNameGenerator.ts)

**Updated deployment scripts:**
- [`00_deploy_bulk_renewal.ts`](https://github.com/ensdomains/ensjs/blob/be91390b/packages/ensjs/deploy/00_deploy_bulk_renewal.ts) - Support both wrapped and unwrapped bulk renewal
- [`00_register_legacy.ts`](https://github.com/ensdomains/ensjs/blob/278b4a69/packages/ensjs/deploy/00_register_legacy.ts) - Fixed type errors, added subname support
- [`00_register_wrapped.ts`](https://github.com/ensdomains/ensjs/blob/be91390b/packages/ensjs/deploy/00_register_wrapped.ts) - Uses wrapped controller

### Test Improvements

**Updated test files:**
- [`commitName.test.ts`](https://github.com/ensdomains/ensjs/blob/6dff985a/packages/ensjs/src/functions/wallet/commitName.test.ts#L58) - Removed debug console.log
- [`registerName.test.ts`](https://github.com/ensdomains/ensjs/blob/6dff985a/packages/ensjs/src/functions/wallet/registerName.test.ts#L33) - Renamed helper for clarity (`getOwner` vs `getNameWrapperOwner`)
- [`renewNames.test.ts`](https://github.com/ensdomains/ensjs/blob/6dff985a/packages/ensjs/src/functions/wallet/renewNames.test.ts) - Removed debug logging
- [`registerHelpers.test.ts`](https://github.com/ensdomains/ensjs/blob/6dff985a/packages/ensjs/src/utils/registerHelpers.test.ts) - Cleaned up, removed tuple tests

**Test contract addresses:**
- Added [`ensDefaultReverseRegistrar`](https://github.com/ensdomains/ensjs/blob/6dff985a/packages/ensjs/src/test/addTestContracts.ts#L33) contract mapping
- Added [`wrappedBulkRenewal`](https://github.com/ensdomains/ensjs/blob/6dff985a/packages/ensjs/src/test/addTestContracts.ts#L100-L102) contract mapping

---

### DNS-over-HTTPS RFC 8484 Compliance

**Problem:** `dnsprovejs` was not following RFC 8484 standard for DNS-over-HTTPS queries, causing failures with some DNS providers.

**Fix:** [`patches/@ensdomains__dnsprovejs@0.5.1.patch`](https://github.com/ensdomains/ensjs/blob/e90c5826/patches/@ensdomains__dnsprovejs@0.5.1.patch)
- Convert base64 to base64url encoding (RFC 4648 §5) - replace `+` with `-`, `/` with `_`, remove `=` padding
- Use proper `Accept: application/dns-message` header instead of `ct` query parameter
- Simplified query string to just `?dns=<base64url>` per RFC 8484

**Related changes:**
- Reverted to `dnsprovejs@0.5.1` (exact version) in [`package.json`](https://github.com/ensdomains/ensjs/blob/e90c5826/packages/ensjs/package.json#L92)
- Applied patch via pnpm patch system

### Contract Naming Corrections

Multiple commits fixed contract naming inconsistencies:
- `NameWrapperPublicResolver` → `WrappedPublicResolver` ([f204ca4b](https://github.com/ensdomains/ensjs/commit/f204ca4b))
- Added missing [`ensDefaultReverseRegistrar`](https://github.com/ensdomains/ensjs/blob/c919189a/packages/ensjs/src/contracts/consts.ts#L21) to types ([c919189a](https://github.com/ensdomains/ensjs/commit/c919189a))

---

### Dependency Updates

**Package updates:**
- `@ensdomains/ens-contracts`: `1.5.2` → `1.6.0` ([package.json](https://github.com/ensdomains/ensjs/blob/6dff985a/packages/ensjs/package.json#L102))
- `hardhat`: `2.24.1` → `^2.22.2` ([package.json](https://github.com/ensdomains/ensjs/blob/6dff985a/packages/ensjs/package.json#L117))
- `hardhat-abi-exporter`: `^2.8.0` → `^2.9.0` ([package.json](https://github.com/ensdomains/ensjs/blob/6dff985a/packages/ensjs/package.json#L118))

**Linting changes:**
- Switched from ESLint to Biome: `lint` command now runs `biome check` ([package.json](https://github.com/ensdomains/ensjs/blob/6dff985a/packages/ensjs/package.json#L79))
- Updated [`biome.json`](https://github.com/ensdomains/ensjs/blob/80a78868/biome.json) configuration
- Fixed linting issues across codebase ([80a78868](https://github.com/ensdomains/ensjs/commit/80a78868))

### TypeScript Configuration

Enhanced TypeScript setup for custom type definitions:
- Updated [`tsconfig.json`](https://github.com/ensdomains/ensjs/blob/80a78868/packages/ensjs/tsconfig.json) to properly handle `@types` directory
- Added type definitions for `dns-packet` and `pako` dependencies
- Improved type safety across the codebase

### Test Environment Patches

**ens-test-env patch** ([`patches/@ensdomains__ens-test-env@1.0.1.patch`](https://github.com/ensdomains/ensjs/blob/4aefe0a2/patches/@ensdomains__ens-test-env@1.0.1.patch)):
- Updated Docker images to `v0.35.0-ens-test-env.2`
- Fixed contract address exports
- Added support for new controller deployments

---

### Project Documentation

**Added [`CLAUDE.md`](https://github.com/ensdomains/ensjs/blob/6dff985a/CLAUDE.md):**
- Comprehensive project overview for AI assistants
- Development commands and testing patterns  
- Architecture documentation
- Import patterns and best practices

---

### For Library Users

**No breaking changes for basic usage.** The main `commitName()` and `registerName()` functions work exactly as before, but now support additional optional parameters:

```typescript
// Before (still works)
await commitName(wallet, {
  name: 'example.eth',
  owner: '0x...',
  duration: 31536000,
  secret: randomSecret(),
})

// After (with new features)
await commitName(wallet, {
  name: 'example.eth',
  owner: '0x...',
  duration: 31536000,
  secret: randomSecret(),
  records: {
    texts: [{ key: 'com.twitter', value: '@example' }],
    coins: [{ coin: 60, value: '0x...' }],
  },
  reverseRecord: ReverseRecordParameter.Default,
  referrer: '0x1234...', // track registration source
})
```

### Breaking Changes

**Removed exports** (unused in ens-app-v3, safe to remove):
- `CommitmentTuple` type
- `RegistrationTuple` type  
- `makeCommitmentTuple()` function
- `makeRegistrationTuple()` function
- `makeCommitmentFromTuple()` function
- `legacyCommitName()` function
- `legacyRegisterName()` function

**If you were using these:** Use the main functions with `RegistrationParameters` instead.

### For Contributors

**New contract structure:**
- `ensEthRegistrarController` = New unwrapped controller (default for registrations)
- `wrappedEthRegistrarController` = NameWrapper-based NFT names
- `legacyEthRegistrarController` = Previous controller (for existing names)

**Testing:**
- Use deployment scripts in `deploy/` directory for comprehensive test scenarios
- `referrableNameGenerator.ts` for testing referral tracking
- All tests pass with new contracts

---

## ✅ Verification Checklist

- [x] All tests pass with new contract addresses
- [x] No external dependencies broken (ens-app-v3 checked)
- [x] Removed code confirmed unused in both ensjs and ens-app-v3
- [x] Contract addresses verified on Etherscan (mainnet & Sepolia)
- [x] DNS-over-HTTPS fix tested with multiple providers
- [x] Backward compatibility maintained for existing usage patterns
- [x] TypeScript types properly exported
- [x] Documentation updated (CLAUDE.md)

---

## 📊 Impact Summary

**Lines changed:** +3,723 / -1,192 across 61 files

**Core changes:**
- ✨ 6 new contract addresses (unwrapped + wrapped controllers)  
- ✨ Enhanced registration with records + referrals
- 🗑️ 458 lines of legacy code removed
- 🐛 DNS RFC compliance fix
- 📚 Comprehensive documentation added

**Testing:**
- ✅ All existing tests pass
- ✅ New deployment scripts for edge cases
- ✅ Test coverage maintained

---

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/ensjs/src/utils/registerHelpers.ts [hunk 1]
```diff
@@ -93,21 +87,18 @@ export const randomSecret = ({
...
   duration,
   resolverAddress = EMPTY_ADDRESS,
-  records: { coins = [], ...records } = { texts: [], coins: [] },
+  records: { coins = [], ...records } = { coins: [], texts: [] },
   reverseRecord,
-  fuses,
   secret,
-}: RegistrationParameters): CommitmentTuple => {
-  const labelHash = labelhash(name.split('.')[0])
+  referrer = '0x0000000000000000000000000000000000000000000000000000000000000000',
```
Comment (LeonmanRolls) [inline packages/ensjs/src/utils/registerHelpers.ts]: Use constant
